Runner fetches Command data from CAS for tasks from Bazel executerequests

dgdonut opened the pull request:
To support this, these structural changes were made:
* Runners support multiple RunTypes (Scoot and Bazel). Runners now keep track of a map to Filers of supported RunTypes. These are set up in main.go's/ICE. It is up to the runner to decide how a runner.Command is of a certain RunType - for now we look at the SnapshotID.
* The filerMap maps to Filers and InitDoneChs (concept taken from closed source workerserver/main.go), and the QueueController can handle Init and Update of the multiple filers (even if in practice we don't expect to use this for bazel)

BzFiler (snapshot.Filer implementation for bazel requests) changed to:
* expose ServerAddr, which identifies the CAS, at top level. required for bazel requests to work.

Invoker detects bazel-type requests and does required preprocessing:
* Fetch the remote exec Command data (where the real Argv & Env live) for the runner.cmd prior to exec'ing it. Tasks started from bazel requests include a digest of the Command data, and the invoker makes a CAS Read request to fetch it. This is done with a CAS read library function as fs_util is designed mainly to handle file operations (and Scoot may eventually want its own CAS functions).

Note:
* Because of a recent /vendor update, you will have to regenerate your dependencies

Commit:
WIP - support multiple filers for RunTypes in queue controller

Commit:
Merge branch 'master' into dgassaway/dx-2705/bzexec

Commit:
fix swarm/recovery tests

Commit:
runner/invoke - bazel executerequests fetch command data from CAS before execing

Commit:
add mockgen comment

ryancouto commented on bazel/cas/client.go:
```diff
@@ -0,0 +1,68 @@
+	}
+	defer cc.Close()
+
+	rname, err := GetReadResourceName("", d.GetHash(), d.GetSizeBytes(), "")
```
Would be nice to have a GetReadResourceNameDefault that doesn't take an instance or fname

dgdonut replied:
done

ryancouto commented on bazel/cas/client.go:
```diff
@@ -0,0 +1,68 @@
+	}
+	req := &bytestream.ReadRequest{
+		ResourceName: rname,
+		ReadOffset:   0,
```
Don't need to specify, 0 will be the default value. Think a constant would be good here though

ryancouto commented on bazel/cas/client.go:
```diff
@@ -0,0 +1,68 @@
+		return nil, fmt.Errorf("Failed to make Read request: %s", err)
+	}
+
+	// Recv from seerver until Limit reached
```
*server

ryancouto commented on bazel/cas/client.go:
```diff
@@ -0,0 +1,68 @@
+// Read data as bytes from a CAS. Takes address in "host:port" format and a bazel Digest to read.
+// Scoot only supports CAS API over a network grpc interface (i.e. an apiserver)
+func ByteStreamRead(serverAddr string, d *remoteexecution.Digest) ([]byte, error) {
+	cc, err := grpc.DialContext(context.Background(), serverAddr)
```
It looks like we aren't doing anything with context.Background(), why not use grpc.Dial(serverAddr)?

dgdonut replied:
yep, done

jasonbs10 commented on bazel/cas/client_test.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+func TestClientRead(t *testing.T) {
+	f := &store.FakeStore{}
```
It's not immediately clear what this is used for.

dgdonut replied:
this was useless; removed

Commit:
bazel request workflow improvements

Commit:
util tool for uploading commands to cas

dgdonut commented:
Other updates:
* got rid of the store.StoreAddr and split logic in workerserver/main.go - I was trying to reuse the http_addr, but needed separate functionality to get the grpc_addr
* added a tool for posting data to CAS - will include in readme an example of how this is used in end-to-end testing
* generally tested working to the limitations of fs_util. will update some error messages/logs

Commit:
document state of end-to-end test, move bazel invoker logic to separate file, minor refactors, error & log standardizations

Commit:
BzFiler uses a Resolver and not hardcoded address. Additional comments in BzFiler and workerserver construction for clarity

dgdonut commented:
This should be considered 'done', requesting further review or approval

jasonbs10 commented on snapshot/bazel/setup.go:
```diff
@@ -13,7 +14,7 @@ func Module() ice.Module {
 	b.PutMany(
 		func() *BzFiler {
-			return MakeBzFilerWithOptions()
+			return MakeBzFiler(dialer.NewConstantResolver(""))
```
Doesn't seem like a useful default.

jasonbs10 commented on bazel/README.md:
```diff
@@ -79,3 +79,26 @@ response {
+INFO[0000] Wrote to CAS successfully                     file:line="bzutil/main.go:73"
+00c1d37dc9582a81e9b4ee2ab0f90e51f28044cd2401507927a0d4ef36310989 28
+
+grpc_cli call localhost:12100 google.bytestream.ByteStream.Write "resource_name: 'uploads/123e4567-e89b-12d3-a456-426655440000/blobs/e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855/0', write_offset: 0, finish_write: true, data: ''"
```
Does this actually accomplish #3 above?

dgdonut replied:
yes, the hand of grpc/protobuf is merciful and kind. I might update this later to data: ' ' (nothing w/ length 1) as it still works, and certain underlying internal blob stores do not accept 0 length

illicitonion replied:
FWIW as of https://github.com/pantsbuild/pants/pull/5378

`fs_util --local-store-path=/Users/dwagnerhall/tmp/store --server-address=localhost:12100 directory save --root=/ ''`

should work. I can put together a pre-release binary if you want to try it out without waiting for the pants release on Friday :)

dgdonut replied:
oh, by certain underlying stores I mean Blobstore :) gives us an HTTP 411 for doing that

jasonbs10 commented on runner/runners/bazel.go:
```diff
@@ -0,0 +1,50 @@
+	digest := cmd.ExecuteRequest.Request.GetAction().GetCommandDigest()
+
+	// Read command data from CAS refernced by filer
+	bzFiler, ok := f.(*bzsnapshot.BzFiler)
```
nit - should this happen in the caller instead?

dgdonut replied:
makes sense, will move

jasonbs10 commented on runner/runners/bazel.go:
```diff
@@ -0,0 +1,50 @@
+	}
+
+	// Update cmd's Argv and EnvVars (overwrite) from Command
+	cmd.Argv = bzCommand.GetArguments()
```
How does the rest of cmd get filled, ex: priority?

dgdonut replied:
The handler in bazel/execution/service.go:execReqToScoot does this. The only things unavailable at that level are the argv/env vars. Adding comments to execReqToScoot

jasonbs10 commented on bazel/cas/client.go:
```diff
@@ -0,0 +1,120 @@
+// These are more straightforward than the server APIs - bytestream provides
+// the majority of the CAS Client implementation. We provide wrappers as
+// higher-level operations.
+// Scoot only supports CAS API over a network grpc interface (i.e. an apiserver)
```
Is there another method that users would expect?

dgdonut replied:
removing. I guess I had in mind some of the other scoot semantics for Stores that try to transparently support local alternatives

Commit:
fix integration test

ryancouto commented on runner/runners/invoke.go:
```diff
@@ -78,6 +77,27 @@ func (inv *Invoker) run(cmd *runner.Command, id runner.RunID, abortCh chan struc
+	if err := bazel.ValidateID(cmd.SnapshotID); err == nil {
+		runType = runner.RunTypeBazel
+	} else {
+		runType = runner.RunTypeScoot
```
I wonder if it would make sense to also validate the ID in the case of a standard Scoot snapshot

ryancouto commented on runner/runners/queue.go:
```diff
@@ -85,36 +86,58 @@ func NewQueueRunner(
+		}
+	}
+
+	if wait {
```
I think you can just call wg.Wait(), and it will immediately pass if it was never added to (i.e. if wait was never set to true)

dgdonut replied:
This has to be broken up because of the way some of our tests and implementations are - they expect if no IDC was given the controller will be available immediately after the constructor returns (this will fail race detector tests). We could make those implementations and/or tests cleaner to avoid this

ryancouto commented on runner/runners/queue.go:
```diff
@@ -85,36 +86,58 @@ func NewQueueRunner(
-			err = <-idc
-			if err != nil {
+			wg.Wait()
+			if initErr != nil {
```
Will this work if we have multiple filers with IDC's? E.g. the first x to return error, and the last to return is successful, then I think initErr will be nil.

dgdonut replied:
initErr only gets set if the IDC returned a non-nil error, so although it can get overwritten, it can't get overwritten with nil (and we log all the individual failures)

ryancouto replied:
ah yep you're right - missed that

Commit:
snapshot/bazel - rework bazel filer based on requirements and safety

ryancouto commented on snapshot/bazel/fs_util_test.go:
```diff
@@ -10,29 +10,26 @@ import (
-	"github.com/twitter/scoot/os/temp"
 )
 
+// uses test vars and setup/teardown defined un filer_test.go
```
*in

Commit:
comment typo

jasonbs10 commented on snapshot/bazel/filer.go:
```diff
@@ -1,50 +1,48 @@
-// Satisfies snapshot.Checkouter, snapshot.Ingester, and snapshot.Updater
-// Default command is fs_util, a tool provided by github.com/pantsbuild/pants which
+// Implements Snapshot interface (snapshot.Checkouter, snapshot.Ingester, and snapshot.Updater)
+// Default runner is fs_util, a tool provided by github.com/pantsbuild/pants which
```
Truncated

jasonbs10 commented on bazel/cas/resource.go:
```diff
@@ -27,6 +27,26 @@ func (r *Resource) String() string {
+		rname += fmt.Sprintf("%s/", instance)
+	}
+	rname += fmt.Sprintf("%s/%s/%d", ResourceNameType, hash, size)
+	if fname != "" {
```
Remind me, what is fname for and why is it optional?

dgdonut replied:
part of the Bazel API spec - described in remote_execution.proto in the description of the ContentAddressableStore service. there is a fair amount of subtlety described in the API this way  https://github.com/googleapis/googleapis/blob/master/google/devtools/remoteexecution/v1test/remote_execution.proto

jasonbs10 replied:
Ugh, I'm not a fan of this api. Could you elaborate a little in the github comments about our interpretation of: 

"the client MAY reuse the `uuid` for uploading
// different blobs. The `resource_name` may optionally have a trailing filename
// (or other metadata) for a client to use if it is storing URLs, as in
// `{instance}/uploads/{uuid}/blobs/{hash}/{size}/foo/bar/baz.cc`"

First of all wtf would you reuse a uuid for a different blob, and what does the server do with the URL fragment? Is it just there for logging purposes?

dgdonut replied:
Yeah, I don't totally understand the point of the fname fragments as there is no prescribed behavior on how to use it on the server - perhaps just for logging as you indicated. Will add a comment on uuid

jasonbs10 commented on binaries/bzutil/main.go:
```diff
@@ -0,0 +1,75 @@
+package main
+
+// Utility functions related to Bazel operations in Scoot.
```
This doesn't seem like an accurate description.

dgdonut replied:
updating

jasonbs10 commented on snapshot/bazel/filer.go:
```diff
@@ -1,44 +1,65 @@
-	command       bzRunner
-	updater       snapshot.Updater
+type BzFiler struct {
+	runner bzRunner
```
A little late for this comment, but why is this considered a runner?

dgdonut replied:
runs underlying bazel operations... I'm open to other suggestions here, this would be easy to change. BzTreeer? BzFs? etc

jasonbs10 replied:
Seems confusing to reuse the term runner, anything else would be better as long as it's unique or thematically consistent.

jasonbs10 commented on runner/runners/queue.go:
```diff
@@ -326,7 +379,9 @@ func (c *QueueController) loop() {
 				c.reqCh = nil
-				c.cancelTimerCh <- nil
+				for i := 0; i < len(c.filerMap); i++ {
+					c.cancelTimerCh <- nil
```
Can't you just close the ch once here?

dgdonut replied:
We start N updateTickers in separate goroutines, not sure how if we recv on one the others will get the message?

jasonbs10 replied:
All the goroutines will receive "_, notOK := <-ch" when a channel is closed afaik.

dgdonut replied:
Ah yes - thanks

jasonbs10 commented on runner/runners/queue.go:
```diff
@@ -123,16 +146,18 @@ func NewQueueRunner(
-	filer         snapshot.Filer
+	filerMap      runner.RunTypeMap
+	updateReq     map[runner.RunType]bool
+	updateLock    sync.Mutex
```
All this locking stuff seems unwieldy, I can't easily tell if it's necessary. It seems like you can just use yet another channel and send it the run type every time the respective ticker updates.

dgdonut replied:
I'm using the lock only when tickers update the map of bool flags that indicate what filers have update requests, and then once in the loop to read & reset the flags.  To queue these up on a channel it would have to be buffered -- to what length? how do updateTickers know if they have a request sitting on the channel or not without draining it? didn't see a clear way of doing that

jasonbs10 replied:
Why does it have to be buffered? They end up being processed serially right? Seems like we can do a blocking send and ultimately it doesn't matter if it gets backed up or not, although I'd expect it shouldn't with reasonable ticker times.

dgdonut replied:
I think an updater could get starved out in this case, if the request channel has a request and before it's drained by the updater (some task is running and taking a while), several updaters hit their tick again and compete to make the next request. Possible with sane values

jasonbs10 replied:
True. In that case I'd use a single ticker with the gcd of filer intervals and then on each tick loop through the filers seeing who all deserves a turn.

jasonbs10 replied:
Up to  you though, whatever you think will be the most understandable.

jasonbs10 reviewed (approved):
Looks like mostly plumbing and docs, so it seems ok as long as you're satisfied with the testing.

Commit:
bazel - don't use gopath for fs_util. clearer bazel task command argv placeholder and logging

Commit:
comments; move bzFiler type assertion for fetching bazel command

Commit:
minor changes

Commit:
rename bzRunner to bzTree. Add stderr to bzCommand error messages

dgdonut committed:
Merge branch 'master' into dgassaway/dx-2705/bzexec
